_posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown

```diff
@@ -0,0 +1,311 @@
+
+    final boolean isAGoodBoy(final Dog dog) {
+        return true;
+    }
+
+}
+{% endhighlight %}
+
+Since this class is `final`, we should use a decorator pattern to extend its' capabilities.
+Let us implement one and add a new `isABadBoy` method:
+
+{% highlight java %}
+final class EvenBetterDogValidator {
+
+    private final DogValidator dogValidator;
+
+    EvenBetterDogValidator(final DogValidator dogValidator) {
+        this.dogValidator = Objects.requireNonNull(dogValidator);
+    }
+
+    final boolean isAGoodBoy(final Dog dog) {
+        return dogValidator.isAGoodBoy(dog);
+    }
+
+    final boolean isABadBoy(final Dog dog) {
+        return !isAGoodBoy(dog);
+    }
+
+}
+{% endhighlight %}
+
+This is a very simple example, but in the real world the decorated classes usually have a multitude of methods.
+Would it not be better if we could just assign a Method Reference to `isAGoodBoy`?
+
+{% highlight java %}
+final class EvenBetterDogValidator {
+
+    private final DogValidator dogValidator;
+
+    EvenBetterDogValidator(final DogValidator dogValidator) {
+        this.dogValidator = Objects.requireNonNull(dogValidator);
+    }
+
+    final boolean isAGoodBoy(final Dog dog) = dogValidator::isAGoodBoy;
+
+    final boolean isABadBoy(final Dog dog) {
+        return !isAGoodBoy(dog);
+    }
+
+}
+{% endhighlight %}
+
+Combining it with the ability to modify Method References with static methods, we can further simplify our code:
+
+{% highlight java %}
+final class EvenBetterDogValidator {
+
+    private final DogValidator dogValidator;
+
+    EvenBetterDogValidator(final DogValidator dogValidator) {
+        this.dogValidator = Objects.requireNonNull(dogValidator);
+    }
+
+    final boolean isAGoodBoy(final Dog dog) = dogValidator::isAGoodBoy;
+
+    final boolean isABadBoy(final Dog dog) = this:not:isAGoodBoy;
+
+}
+{% endhighlight %}
+
+Even the method parameter names could be made optional for the same benefits I described before: fewer things to name.
+
+A similar feature is already available in Kotlin, called [single-expression functions](https://kotlinlang.org/docs/reference/functions.html#single-expression-functions).
+Although they cannot accept method references, these functions can significantly reduce the amount of boilerplate.
+This is how our decorator class could look like if it were written in Kotlin: 
+
+{% highlight kotlin %}
+class EvenBetterDogValidator(private val dogValidator: DogValidator) {
+
+    fun isAGoodBoy(dog: Dog) = dogValidator.isAGoodBoy(dog);
+
+    fun isABadBoy(dog: Dog) = !isAGoodBoy(dog);
+
+}
+{% endhighlight %}
+
+## Summary
+
+I recommend using Method References whenever possible.
+Your code will become cleaner and easier to refactor, at no extra cost.
+Keep in mind though that they have some limitations and unless those get waived in the next Java releases, some workarounds need to be implemented.
+
+Method references are—in my humble opinion—not a complete feature.
```

Comment:
```suggestion
Method references are - in my humble opinion - not a complete feature.
```

Reply:
These are em dashes though!

Reply:
interesting... I didn't know about them. Maybe that's a sign they're not universal enough? 😉